proposals/nnnn-structured-concurrency.md

```diff
@@ -377,13 +377,43 @@ Much of the proposed implementation of structured concurrency is in the APIs for
 
 #### `Task` type
 
-The `Task` type is an empty `enum` that is used as a namespace for types and operations related to task management.
+The `Task` type is used as a namespace for types and operations related to task management. 
+
+```swift
+struct Task: Equatable, Hashable { ... }
+```
+
+The "current task" can be obtained from inside of an async function using the `current` static function, like this:
+
+```swift
+func foo() async { 
+  let task = await Task.current()
+}
+```
+
+An asynchronous function _always_ executes within a task. As the `current` function itself is also asynchronous, it can only be invoked from an asynchronous function and is guaranteed to return the apropriate `Task`. The task object obtained this way is safe access by other tasks/threads. The APIs surfaced on it are specifically designed to also be safe, and relatively cheap, to be called from other tasks.
+
+Tasks are equatable and hashable, this can be used to store and compare them, and can be used e.g. to answer questions such as "is this the same task I was called from before" etc. Keep in mind though, that tasks should not be held onto unnecessarily.
+
+#### `UnsafeCurrentTask` type
+
+It is also possible to obtain a reference to the "current" task from _synchronous_ functions.
+
+This type is the core underlying type used to implement most of the APIs discussed in the following sections (which may be used from non-async contexts). Accessing using this API performs a thread-local lookup of a specific thread-local variable that is maintained by the Swift concurrency runtime.
+
+The `UnsafeCurrentTask` is unsafe because it offers APIs which must not be invoked from any other task than the task it represents. In other words, it must not be stored, or accessed from other tasks/threads. Invoking some of its APIs from other contexts will result in undefined behavior.
 
 ```swift
-@frozen enum Task { }
+func synchronous() {
+  if let task = Task.unsafeCurrent {
```

Comment:
> Also, can we get a Task from an UnsafeCurrentTask? (It seems like we should be able to). 

This seems reasonable; as Task has only "safe" APIs it makes sense to get it from an `UnsafeCurrentTask` 🤔 

Will amend this

> Can we get a Task from a Task.Handle? (It seems like we should be able to). 

I don't think that should be the case actually, but I can be convinced otherwise. It's not a core property and need of a "read side" of the work (the Task.Handle) to be able to query stuff about the `Task`. I know people "asked for that on the forums" but what's the actual reason? Why would a _read side_ ever query isCanceled or priority of ongoing work -- all it needs is to wait for the handle to complete, nothing else.

> In this case, we effectively have three copies of those Task APIs that are safe to use from effectively anywhere

Reply:
`Task.Handle` already has `isCancelled` and `cancel`. If you're holding a handle for some work, you might want to cancel it (if you no longer care about the result) or determine whether it's been cancelled (say, to drop that handle from the set of results you're still waiting for).

Reply:
Okey, lemme add a `handle.task` then 👍